### PR TITLE
Consolidate all initialization in a single Service

### LINKED ISSUE #1683: Move route subscription out of app component
**Describe the bug**
https://github.com/DSpace/dspace-angular/issues/1668 reported route dependent themes wouldn't be resolved correctly on first page load. This was fixed in https://github.com/DSpace/dspace-angular/pull/1669 by running that theme check on a later route event, if it hadn't already happened.

However that's a workaround. The reason for the issue seems to be that we subscribe to the route events inside a component, and when that component is first rendered, most of the events for the first route operation have already happened. So a better solution would be to move that subscription out of the component

**Expected behavior**
We should subscribe to the route in a single place, early enough to catch all route events, so that can't happen inside a component. A dedicated service, or an `APP_INITIALIZER` function is likely a better idea.

Afterwards we should roll back the changes in https://github.com/DSpace/dspace-angular/pull/1669

**Related work**
- https://github.com/DSpace/dspace-angular/issues/1668
- https://github.com/DSpace/dspace-angular/pull/1669

### PR DESCRIPTION
**Pull request by ybnd**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #1683
* More robust approach to address #1668

## Description

This PR moves most app initialization to a new `InitService` in order to

* make sure route-listening initialization methods are called as early as possible
  * After #1627, `AppComponent` does not receive the earliest Router events, which lead to issues with route-dependent themes
* have just one `APP_INITIALIZER` provider
* keep all initialization logic close together and more neatly separated between server/browser
  * before, initialization was split between `AppModule`, `*AppModule`, `*DSpaceStateTransfer` and `AppComponent`
  * `AppComponent` used `@Optional` to skip some browser-specific steps on the server

<br>

Some extra context

* A side effect of injecting `Router` into an `APP_INITIALIZER` dependency was that now `APP_BASE_HREF` _must_ be fully resolved by the time `InitService.init()` is called, otherwise the app crashes during the SSR→CSR transition. 
  To address ths we moved the `extendEnvironmentWithAppConfig()` call to a "pre-initialization step" to ensure it is called _before_ `init()`.

* The providers required for `InitService` are closely linked to it and a bit finnicky, so we thought it best _not_ to duplicate them in `ServerAppModule`/`BrowserAppModule`. Instead, they are defined once in `InitService.providers()`.

<br>

Going forward, when adding new initialization logic:

* Check if it can actually be done in `*InitService`
  * Services that (indirectly) depend on `ApplicationRef` can only be injected _after_ `APP_INITIALIZER`, so they cannot be used in `InitService` (e.g. `NgbModal`)
  * Anything that needs to interact with a component template can't be moved from that component
* Add a new method for this particular initialization step. If it can be reused between the server and browser implementations, add it to `InitService`.
* Call this method from `ServerInitService` and/or `BrowserInitService`

## Instructions for Reviewers
Confirm that the app keeps working in exactly the same way as before this PR.

Confirm that #1668 doesn't reappear.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Review by tdonohue — COMMENTED** [review]
@ybnd : This looks good overall, and I'm basically a +1.  I just have some very minor requests inline below (mostly for TypeDocs/comments). The code itself looks good...and I tried to review it thoroughly to make sure nothing was lost/overlooked in the reorganization (I didn't notice anything).

I also tested this in prod & dev mode (both work). I tried the `custom` theme (works) and made sure Google Analytics & Klaro still worked (both do).  Basically tried to test as many of the pieces that are loaded on initialization as I could...and I haven't found any obvious bugs.

Once my minor feedback is addressed, I can be considered a +1.  So, once @atarix83 approves, then this can be merged.

File: src/app/core/shared/distinct-next.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*
+ * something something atmire
```
[the remark this answers is not in the document]
Comment (ybnd) [reply src/app/core/shared/distinct-next.ts:2]: woops

**Review by tdonohue — APPROVED** [review]
👍 Thanks @ybnd !  This looks great to me now.  All my previously requested (minor) feedback has been addressed.

@atarix83 : when you get a chance, it'd be good to get your feedback on this PR so that we can move it forward quickly.  Thanks!

**Review by atarix83 — APPROVED** [review]
thanks @ybnd  for this PR

I've overall tested the application wihtout finding any problems. I've also checked the mentioned issue is resolved

**Pull request merged**